staging/src/k8s.io/apiserver/pkg/endpoints/installer.go

```diff
@@ -68,6 +68,72 @@ type action struct {
 	AllNamespaces bool // true iff the action is namespaced but works on aggregate result for all namespaces
 }
 
+func ConvertResourceInfoToDiscovery(resources []*apiResourceInfo) ([]metav1.APIResource, []metav1.APIResourceDiscovery) {
+	var legacyResources []metav1.APIResource
+	for _, r := range resources {
+		// apiResourceInfo.APIResource maps 1:1 with the legacy resource types
+		legacyResources = append(legacyResources, r.APIResource)
+	}
+
+	var apiResourceList []metav1.APIResourceDiscovery
+	rootResources := map[schema.GroupVersionResource]*metav1.APIResourceDiscovery{}
+
+	// Loop through all top-level resources
+	for _, r := range resources {
+		if len(r.Subresource) > 0 {
+			// Skip subresources for now so we can get the list of resources
+			continue
+		}
+
+		var scope metav1.ResourceScope
+		if r.Namespaced {
+			scope = metav1.ScopeNamespace
+		} else {
+			scope = metav1.ScopeCluster
+		}
+
+		
+		apiResourceList = append(apiResourceList, metav1.APIResourceDiscovery{
+			Resource: r.Name,
+			Scope:    scope,
+			ReturnType: metav1.APIDiscoveryKind{
+				Group:   r.Group,
+				Version: r.Version,
+				Kind:    r.Kind,
+			},
+			Verbs:      r.Verbs,
+			ShortNames: r.ShortNames,
+			Categories: r.Categories,
+		})
+		rootResources[r.Resource] = &apiResourceList[len(apiResourceList) - 1]
+	}
+
+	// Loop through all subrouces
+	for _, r := range resources {
+		if len(r.Subresource) == 0 {
+			// Skip root resources
+			continue
+		}
+
+		parent, exists := rootResources[r.Resource]
+		if !exists {
+			// Had a subresource but not its parent???
+			continue
+		}
+
+		parent.Subresources = append(parent.Subresources, metav1.APISubresourceDiscovery{
+			Subresource: r.Subresource,
+			ReturnType: &metav1.APIDiscoveryKind{
```

Comment:
I need to chase this back  through.  Not all subresources have a return type. For instance, pods/exec.